### PR TITLE
now support Date#step, Date#upto, Date#downto

### PR DESCRIPTION
**Pull request by bwl21**
see #1694

I have implemented these methods in pure ruby. Maybe there is already a pure ruby implementation out, but I was not able to find it.

### CODE REVIEW

File: stdlib/date.rb [hunk 1]
```diff
@@ -596,6 +598,40 @@ def tuesday?
...
   end
 
+  def step(limit, step = 1, &block)
+    steps_count = (limit - self).to_i
+
+    if (steps_count * step) < 0
+      steps = []
+    else
+      if steps_count < 0
+        puts "steps_count -: #{steps_count}"
```
Comment (iliabylich) [inline stdlib/date.rb]: Could you remove this debug code?

File: stdlib/date.rb [hunk 2]
```diff
@@ -596,6 +598,40 @@ def tuesday?
...
+        steps = (0 .. -steps_count).step(step.abs).map {|i| -i}.reverse
```
Comment (iliabylich) [inline stdlib/date.rb]: Please,
1. Remove space from the range definition (i.e. `(0..-steps_count)`
2. Add spaces around curly braces (i.e. `.map { |i| -i }`

File: stdlib/date.rb [hunk 3]
```diff
@@ -596,6 +598,40 @@ def tuesday?
...
+      else
+        puts "steps_count +: #{steps_count}"
+        steps = (0 .. steps_count).step(step.abs)
+      end
+    end
+
+    result = steps.map {|i| self + i}
```
Comment (iliabylich) [inline stdlib/date.rb]: Missing spaces around `{` `}`

File: stdlib/date.rb [hunk 4]
```diff
@@ -596,6 +598,40 @@ def tuesday?
...
+    if block_given?
+      result.each {|i| yield(i)}
+      self
+    else
+      result
+    end
+  end
+
+  def upto(enddate, &block)
+       self.step(enddate, 1, &block)
```
Comment (iliabylich) [inline stdlib/date.rb]: Please, remove redundant `self` usage and align it properly (2 spaces)

**Comment by bwl21** [comment]
Thanks for review. I will edit as requested

**Comment by iliabylich** [comment]
@bwl21 There's still one more `puts` in your patch.

**Comment by iliabylich** [comment]
@bwl21 Thank you

**Pull request merged**